client/lib/user/support-user-interop.js

```diff
@@ -18,27 +24,117 @@ import { getSupportUser, getSupportToken } from 'state/support/selectors';
  *
  * @param {Object}  reduxStore  The global redux store instance
  */
-export default function( reduxStore ) {
-	const user = new User();
+class SupportUser {
+	constructor() {
+		debug( 'Support User is enabled in this environment' );
 
-	// Called when the support user token was updated in wpcom
-	const onTokenChange = () => {
-		user.fetch();
-		userSettings.fetchSettings();
+		this.reduxStoreReady = new Promise( ( resolve ) => {
+			this.setReduxStore = ( reduxStore ) => resolve( reduxStore );
+		} );
 	}
 
-	// Called when an API call fails due to a token error
-	const onTokenError = ( error ) => {
-		reduxStore.dispatch( supportUserRestore( error ) );
+	fetchToken( user, password ) {
+		debug( 'Fetching support user token' );
+
+		return this.reduxStoreReady.then( ( reduxStore ) => {
+			reduxStore.dispatch( supportUserTokenFetch( user ) );
+
+			const setToken = ( response ) => {
+				this.rebootWithToken( response.username, response.token );
+			};
+
+			const errorFetchingToken = ( error ) => {
+				reduxStore.dispatch( supportUserError( error.message ) );
+			};
+
+			return wpcom.fetchSupportUserToken( user, password )
+				.then( setToken )
+				.catch( errorFetchingToken );
+		} );
 	}
 
-	// Follow any changes to the support token in the Redux store and
-	// update the wpcom API interceptor accordingly.
-	reduxStore.subscribe( () => {
-		const state = reduxStore.getState();
+	/**
+	 * Reboot normally as the main user
+	 */
+	rebootNormally() {
+		debug( 'Rebooting Calypso normally' );
 
-		if ( wpcom.setSupportUserToken( getSupportUser( state ), getSupportToken( state ) ) ) {
-			onTokenChange();
+		store.remove( STORAGE_KEY );
+		window.location.reload();
+	}
+
+	/**
+	  * Reboot Calypso as the support user
+	  * @param  {string} user  The support user's username
+	  * @param  {string} token The support token
+	  */
+	rebootWithToken( user, token ) {
+		debug( 'Rebooting Calypso with support user' );
+
+		store.set( STORAGE_KEY, { user, token } );
+		window.location.reload();
```

Comment:
Good idea. @aduth can you think of any other cases we might run into besides login/logout/switch user where current user id is updated?